full-service/src/util/b58/mod.rs

```diff
@@ -0,0 +1,211 @@
+//! Public Address base58 encoding and decoding.
+
+pub mod errors;
+mod tests;
+pub use self::errors::B58Error;
+
+use bip39::{Language, Mnemonic};
+use mc_account_keys::{AccountKey, PublicAddress, RootEntropy, RootIdentity};
+use mc_account_keys_slip10::Slip10KeyGenerator;
+use mc_crypto_keys::CompressedRistrettoPublic;
+use serde::{Deserialize, Serialize};
+use std::convert::TryFrom;
+
+pub struct DecodedPaymentRequest {
+    pub public_address: PublicAddress,
+    pub value: u64,
+    pub memo: String,
+}
+
+pub struct DecodedTransferPayload {
+    pub root_entropy: Option<RootEntropy>,
+    pub bip39_entropy: Option<Vec<u8>>,
+    pub account_key: AccountKey,
+    pub txo_public_key: CompressedRistrettoPublic,
+    pub memo: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub enum PrintableWrapperType {
+    PublicAddress,
+    PaymentRequest,
+    TransferPayload,
+}
+
+pub fn b58_printable_wrapper_type(b58_code: String) -> Result<PrintableWrapperType, B58Error> {
+    let wrapper = mc_mobilecoind_api::printable::PrintableWrapper::b58_decode(b58_code)?;
+
+    if wrapper.has_payment_request() {
+        return Ok(PrintableWrapperType::PaymentRequest);
+    } else if wrapper.has_transfer_payload() {
+        return Ok(PrintableWrapperType::TransferPayload);
+    } else if wrapper.has_public_address() {
+        return Ok(PrintableWrapperType::PublicAddress);
+    }
+
+    Err(B58Error::NotPrintableWrapper)
+}
+
+pub fn b58_encode(public_address: &PublicAddress) -> Result<String, B58Error> {
+    let mut wrapper = mc_mobilecoind_api::printable::PrintableWrapper::new();
+    wrapper.set_public_address(public_address.into());
+    Ok(wrapper.b58_encode()?)
+}
+
+pub fn b58_decode(b58_public_address: &str) -> Result<PublicAddress, B58Error> {
+    let wrapper = mc_mobilecoind_api::printable::PrintableWrapper::b58_decode(
+        b58_public_address.to_string(),
+    )?;
+
+    let pubaddr_proto: &mc_api::external::PublicAddress = if wrapper.has_payment_request() {
+        let payment_request = wrapper.get_payment_request();
+        payment_request.get_public_address()
+    } else if wrapper.has_public_address() {
+        wrapper.get_public_address()
+    } else {
+        return Err(B58Error::NotPublicAddress);
+    };
+
+    let public_address = PublicAddress::try_from(pubaddr_proto)?;
+    Ok(public_address)
+}
+
+pub fn decode_public_address_b58(public_address_b58_code: &str) -> Result<PublicAddress, B58Error> {
```

Comment:
Whats the difference between this and `b58_decode`?
Why would you ever want to use this one?

Reply:
(Is it actually used at all?)